File: server/server_main/app/routes/devices.py
"""Routes consumed by the ESP32 device (not the user browser)."""
from datetime import datetime, timezone
from functools import wraps
from flask import Blueprint, request, jsonify, current_app
from ..extensions import db
from ..models.kiosk import Kiosk
from ..models.device_telemetry import DeviceTelemetry
from ..models.charging_session import ChargingSession
from ..services.command_service import get_pending_commands, ack_command

devices_bp = Blueprint("devices", __name__)


def require_device_key(f):
    """Decorator: validates the device API key from Authorization header."""
    @wraps(f)
    def decorated(*args, **kwargs):
        auth = request.headers.get("Authorization", "")
        token = auth.removeprefix("Bearer ").strip()
        if token != current_app.config["DEVICE_API_KEY"]:
            return jsonify({"error": "unauthorized"}), 401
        return f(*args, **kwargs)
    return decorated


def _resolve_kiosk() -> tuple[Kiosk | None, dict | None]:
    """Read kiosk_id from query string or JSON body, return kiosk or error response."""
    kiosk_id = request.args.get("kiosk_id") or (request.get_json(silent=True) or {}).get("kiosk_id")
    if not kiosk_id:
        return None, ({"error": "kiosk_id is required"}, 400)
    kiosk = Kiosk.query.get(int(kiosk_id))
    if not kiosk:
        return None, ({"error": "kiosk not found"}, 404)
    return kiosk, None


@devices_bp.get("/commands")
@require_device_key
def poll_commands():
    """ESP32 calls this every 2 seconds to get pending commands."""
    kiosk, err = _resolve_kiosk()
    if err:
        return jsonify(err[0]), err[1]

    kiosk.touch()
    db.session.commit()

    pending = get_pending_commands(kiosk.id)
    return jsonify({"commands": [c.to_dict() for c in pending]})


@devices_bp.post("/commands/<int:command_id>/ack")
@require_device_key
def ack(command_id):
    """ESP32 calls this after executing a command."""
    kiosk, err = _resolve_kiosk()
    if err:
        return jsonify(err[0]), err[1]

    cmd = ack_command(command_id, kiosk.id)
    if not cmd:
        return jsonify({"error": "command not found or already acked"}), 404

    db.session.commit()
    return jsonify({"acked": command_id})


@devices_bp.post("/telemetry")
@require_device_key
def receive_telemetry():
    """ESP32 posts sensor readings every 5 seconds."""
    data = request.get_json(silent=True) or {}
    kiosk_id = data.get("kiosk_id")

    if not kiosk_id:
        return jsonify({"error": "kiosk_id is required"}), 400

    kiosk = Kiosk.query.get(int(kiosk_id))
    if not kiosk:
        return jsonify({"error": "kiosk not found"}), 404

    kiosk.touch()

    telemetry = DeviceTelemetry(kiosk_id=kiosk.id)
    telemetry.set_port_data(data.get("ports", []))
    telemetry.bin_level = data.get("bin_level")
    db.session.add(telemetry)

    # Auto-complete charging sessions whose duration has elapsed
    _auto_complete_expired_sessions(kiosk.id)

    db.session.commit()
    return jsonify({"received": True})


def _auto_complete_expired_sessions(kiosk_id: int):
    """Mark charging sessions as completed if their duration has passed."""
    from ..services.command_service import queue_command
    active = ChargingSession.query.filter_by(kiosk_id=kiosk_id, status="active").all()
    now = datetime.now(timezone.utc)
    for session in active:
        elapsed = (now - session.started_at.replace(tzinfo=timezone.utc)).total_seconds()
        if elapsed >= session.duration_seconds:
            session.complete()
            queue_command(kiosk_id, "deactivate_port", {"port": session.port_number})

File: server/server_main/src/routes/devices.ts
import { Router, Response, NextFunction } from 'express'
import { z } from 'zod'
import prisma from '../prisma'
import { requireDeviceKey } from '../middleware/deviceAuth'
import { getPendingCommands, ackCommand, queueCommand } from '../services/commandService'
import { broadcastToAdmin, broadcastToKiosk } from '../services/sseService'
import { getPortStatus } from './kiosk'
import { awardCredits } from '../services/creditService'
import { log } from '../logger'
import { DeviceRequest } from '../types'

const router = Router()
router.use(requireDeviceKey)

async function touchKiosk(kioskId: number) {
  return prisma.kiosk.update({
    where: { id: kioskId },
    data: { status: 'online', lastSeenAt: new Date() },
  })
}

// GET /commands
router.get('/commands', async (req: DeviceRequest, res: Response, next: NextFunction) => {
  try {
    const kioskId = req.kiosk!.id

    await touchKiosk(kioskId)

    const commands = await getPendingCommands(kioskId)
    if (commands.length > 0) {
      log.device(`kiosk #${kioskId} poll — ${commands.length} pending command(s): ${commands.map(c => c.commandType).join(', ')}`)
    }

    res.json({
      commands: commands.map(c => ({
        id:           c.id,
        command_type: c.commandType,
        payload:      c.payload ? (() => { try { return JSON.parse(c.payload!) } catch { return {} } })() : {},
        status:       c.status,
        created_at:   c.createdAt,
        acked_at:     c.ackedAt,
      })),
    })
  } catch (err) {
    next(err)
  }
})

// POST /commands/:id/ack
router.post('/commands/:id/ack', async (req: DeviceRequest, res: Response, next: NextFunction) => {
  try {
    const commandId = parseInt(req.params.id)
    const kioskId   = req.kiosk!.id
    log.device(`kiosk #${kioskId} ACK command #${commandId}`)
    await ackCommand(commandId, kioskId)
    res.json({ acked: commandId })
  } catch (err) {
    next(err)
  }
})

const ultrasonicSchema = z.object({
  entrance_cm: z.number().optional(),
  bin_top_cm:  z.number().optional(),
  bin_bot_cm:  z.number().optional(),
}).optional()

const telemetrySchema = z.object({
  kiosk_id:           z.number().int().positive().optional(),
  ports:              z.array(z.object({
    voltage:   z.number().optional(),
    current:   z.number().optional(),
    relay_on:  z.boolean().optional(),
  })).optional(),
  bin_level:          z.number().int().optional(),
  ultrasonic:         ultrasonicSchema,
  bottle_at_entrance: z.boolean().optional(),
  bottle_in_bin:      z.boolean().optional(),
  fsm_state:          z.string().optional(),
})

// POST /telemetry
router.post('/telemetry', async (req: DeviceRequest, res: Response, next: NextFunction) => {
  try {
    const body = telemetrySchema.parse(req.body)
    const {
      ports     = [],
      bin_level,
      ultrasonic,
      bottle_at_entrance,
      bottle_in_bin,
      fsm_state,
    } = body
    const kioskId = req.kiosk!.id

    await touchKiosk(kioskId)

    await prisma.deviceTelemetry.create({
      data: {
        kioskId,
        portData: JSON.stringify(ports),
        binLevel: bin_level ?? null,
      },
    })

    const activePorts = ports.filter(p => p.relay_on).length
    log.device(
      `kiosk #${kioskId} telemetry — ` +
      `fsm=${fsm_state ?? 'idle'} ` +
      `bin=${bin_level ?? '?'}% ` +
      `ports=${activePorts} active ` +
      `entrance=${bottle_at_entrance ?? false} ` +
      `inBin=${bottle_in_bin ?? false}`,
    )

    // Auto-complete expired charging sessions
    const activeSessions = await prisma.chargingSession.findMany({
      where: { kioskId, status: 'active' },
    })

    for (const session of activeSessions) {
      const elapsed = Math.floor((Date.now() - session.startedAt.getTime()) / 1000)
      if (elapsed >= session.durationSeconds) {
        log.charging(`kiosk #${kioskId} — session #${session.id} expired (port ${session.portNumber}), auto-completing`)
        await prisma.chargingSession.update({
          where: { id: session.id },
          data:  { status: 'completed', endedAt: new Date() },
        })
        await queueCommand(kioskId, 'deactivate_port', { port: session.portNumber })
        log.charging(`kiosk #${kioskId} — queued deactivate_port for port ${session.portNumber}`)
      }
    }

    // ── Bin confirmation — award credits when bottle lands ────────────────────
    if (bottle_in_bin === true) {
      log.device(`[Stage 5] bottle_in_bin=true received from ESP32 — kiosk #${kioskId}, looking for pending deposit`)

      const pendingDeposit = await prisma.bottleDeposit.findFirst({
        where:   { status: 'pending_bin', session: { kioskId } },
        include: { session: true },
        orderBy: { timestamp: 'desc' },
      })

      if (pendingDeposit) {
        log.device(`[Stage 5] Confirming deposit #${pendingDeposit.id} — ${pendingDeposit.creditsAwarded} credits → user #${pendingDeposit.session.userId}`)

        await prisma.bottleDeposit.update({
          where: { id: pendingDeposit.id },
          data:  { status: 'confirmed' },
        })

        if (pendingDeposit.creditsAwarded > 0) {
          await awardCredits(
            pendingDeposit.session.userId,
            pendingDeposit.creditsAwarded,
            'bottle_deposit',
            pendingDeposit.id,
          )
          log.device(`kiosk #${kioskId} — awarded ${pendingDeposit.creditsAwarded} credits to user #${pendingDeposit.session.userId}`)
        }

        broadcastToKiosk(String(kioskId), {
          type:            'bottleInBin',
          confirmed:       true,
          deposit_id:      pendingDeposit.id,
          credits_awarded: pendingDeposit.creditsAwarded,
        })
      } else {
        log.deviceWarn(`kiosk #${kioskId} — bottle_in_bin=true but no pending deposit found`)
      }
    } else if (bottle_in_bin === false && fsm_state === 'confirming') {
      log.device(`kiosk #${kioskId} — bin timeout (fsm=confirming), rejecting pending deposit`)

      const pendingDeposit = await prisma.bottleDeposit.findFirst({
        where:   { status: 'pending_bin', session: { kioskId } },
        orderBy: { timestamp: 'desc' },
      })

      if (pendingDeposit) {
        await prisma.bottleDeposit.update({
          where: { id: pendingDeposit.id },
          data:  { status: 'rejected', creditsAwarded: 0 },
        })
        log.deviceWarn(`kiosk #${kioskId} — deposit #${pendingDeposit.id} rejected (bin timeout)`)
      }

      broadcastToKiosk(String(kioskId), {
        type:            'bottleInBin',
        confirmed:       false,
        credits_awarded: 0,
      })
    }

    // Broadcast telemetry to admin SSE
    broadcastToAdmin({
      type:              'telemetry',
      kioskId,
      ports,
      binLevel:          bin_level ?? null,
      ultrasonic:        ultrasonic ?? null,
      bottleAtEntrance:  bottle_at_entrance ?? false,
      bottleInBin:       bottle_in_bin ?? false,
      fsmState:          fsm_state ?? 'idle',
      timestamp:         new Date(),
    })

    // Broadcast port + FSM state to kiosk SSE clients
    const portStatus = await getPortStatus(kioskId)
    broadcastToKiosk(String(kioskId), {
      type:             'ports',
      ports:            portStatus,
      bottleAtEntrance: bottle_at_entrance ?? false,
      fsmState:         fsm_state ?? 'idle',
    })

    res.json({ received: true })
  } catch (err) {
    next(err)
  }
})

export default router
